### ISSUE TITLE
Suggested map_or_else when value partialy moved

### ISSUE THREAD

**Issue by t-webber** [rust-lang/rust-clippy#13964]
### Summary

The `clippy::map_or_else` lint suggests erroneous code when partially moving the value in one arm and using it in another. 

The ownership is dropped before the test for a `map_or_else`, but only after the test when doing a match.

### Lint Name

map_or_else

### I tried this code

```rust
#![warn(clippy::nursery)]

#[derive(Default, Debug)]
struct A(Option<String>);

fn main() {
    let a = A::default();
    let _ = match a.0 {
        Some(x) => x,
        None => panic!("{a:?} is invalid."),
    };
}

```

### I saw this happen

```
warning: use Option::map_or_else instead of an if let/else
[...]
help: try: `a.0.map_or_else(|| panic!("{a:?} is invalid."), |x| x)`
```

### If I follow the lint

```rust
#[derive(Default, Debug)]
struct A(Option<String>);

fn main() {
    let a = A::default();
    let _ = a.0.map_or_else(|| panic!("{a:?} is invalid."), |x| x);
}
```

I get a compilation error: `borrow of partially moved value: a` (this is normal as `a` is moved when accessing the `0` field).

### I expected

I expected the warning not to be triggered in this case

### Version

```text
rustc 1.85.0-nightly (7f75bfa1a 2024-12-30)
binary: rustc
commit-hash: 7f75bfa1ad4e9a9d33a179a90603001515e91991
commit-date: 2024-12-30
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.6
```

### Additional Labels

@rustbot label +l-suggestion-causes-error
